### PR TITLE
packedjson port

### PR DESCRIPTION
**Pull request by planetis-m**
Port from stdlib's json to Araq/packedjson Not sure of the benefits but seemed like a good exercise :P

### CODE REVIEW

File: src/haloniumpkg/errorhandler.nim [hunk 1]
```diff
@@ -125,24 +125,24 @@ proc checkResponse*(response: JsonNode) =
     value: JsonNode
 
   let
-    isInt = (not status.isNil and status.kind == JInt)
+    isInt = (status.kind != JNull and status.kind == JInt)
 
-  if status.isNil or (isInt and status.getInt() == ErrorCode.Success.int):
+  if status.kind != JNull or (isInt and status.getInt() == ErrorCode.Success.int):
```
Comment (planetis-m) [inline src/haloniumpkg/errorhandler.nim:130]: Ha! found it!

**Comment by planetis-m** [comment]
Needs to merge some changes to packedjson and tag a new release.

File: src/haloniumpkg/webdriver.nim [hunk 2]
```diff
@@ -1645,7 +1642,7 @@ proc findElements*(element: Element, selector: string, strategy = CssSelector):
       getSelectorParams(element.session, selector, strategy),
       stopOnException = false
     )
-    for elementNode in response["value"].to(seq[JsonNode]):
+    for elementNode in response["value"].items:
```
Comment (planetis-m) [inline src/haloniumpkg/webdriver.nim:1645]: Is this correct?

File: src/haloniumpkg/webdriver.nim [hunk 3]
```diff
@@ -573,30 +570,30 @@ proc findElements*(self: Session, selector: string, strategy = CssSelector): seq
       getSelectorParams(self, selector, strategy),
       stopOnException = false
      )
-    for elementNode in response["value"].to(seq[JsonNode]):
+    for elementNode in response["value"].items:
```
Comment (planetis-m) [inline src/haloniumpkg/webdriver.nim:573]: is this correct?

**Comment by planetis-m** [comment]
The reasoning of choosing between `JsonNode/JsonTree` is the following:

1. Variables that can be modified are of type `JsonTree`. Disallow modification by using `JsonNode`.

1. Have every return / output parameter type be `JsonTree` and make sure that you pass copies or fresh `JsonTree`'s.
1. Input parameters should be of type `JsonNode`.

1. Remove unnecessary copies by casting to `JsonTree` when you are sure about it.

**Comment by jyapayne** [comment]
@planetis-m from what I can tell, packedjson hasn't been worked on in quite a while (~3 years) vs the stdlib json module. I'm not sure if the performance difference is that different anymore and I cannot get the benchmark to work in the `packedjson` repo to see if the code works.

Is there a reason besides practice that you wanted to make this change?

**Comment by planetis-m** [comment]
@jyapayne I don't think anything has changed significantly with json, so I expect for the benchmark result to still hold. You can run the tests inside the packedjson, for me they pass. I thought it would improve performance, but I understand that's questionable.

File: src/haloniumpkg/webdriver.nim (hunk 2 above)
Comment (planetis-m) [reply src/haloniumpkg/webdriver.nim:1645]: To answer, yes `to(seq[JsonNode])` only accepts `JArray` node and so does `items`

**Comment by planetis-m** [comment]
Nim 1.0.6 fails to compile it successful. it should work

**Comment by jyapayne** [comment]
In general, I am against this PR. From my point of view, there is no good reason to use `packedjson` other than memory performance, but there aren't any large json objects being used in this repo. And I think it harms useability with regular Nim, since now the user has to learn about differences between JsonNode and JsonTree, and then how those differ from the JsonNode in the stdlib. (Also, with your changes, `examples/uploadfile.nim` is broken)

If you can give me a benchmark for how this change affects memory usage/performance for halonium on a regular use case, I'd be happy to take a look again!

**Comment by planetis-m** [comment]
Users shouldn't be that affected JsonNode can be used uniformly. Except when serializing JsonTree is used but converters to JsonNode make this seamless. But I also added more documentation Araq/packedjson#10 . Keep in mind std / json is also a reference type and has aliasing problems that these two types try to prevent. I will try to run a test, any suggestions?

**Comment by planetis-m** [comment]
For the record having read the packedjson code, I think for hardly nested JSON objects, packedjson can be almost linear time (because it stores length info of atoms, but has to do string comparison for keys), worst than Tables O(1), I know, but for small objects it could faster/less memory used.

**Pull request closed without merging**